README.md
# structural-variantcalling

This repository contains the [Biowdl](https://github.com/biowdl) 
workflow for structural variant calling.

## Documentation

You can find the documentation [here](https://biowdl.github.io/structural-variantcalling/).

## About
This pipeline is part of [Biowdl](https://github.com/biowdl),
developed by [the SASC team](http://sasc.lumc.nl/). 

## Contact

<p>
  <!-- Obscure e-mail address for spammers -->
For any question related to this workflow, please use the
<a href='https://github.com/biowdl/RNA-seq/issues'>github issue tracker</a>
or contact
 <a href='http://sasc.lumc.nl/'>the SASC team</a> directly at: <a href='&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#115;&#97;&#115;&#99;&#64;&#108;&#117;&#109;&#99;&#46;&#110;&#108;'>
&#115;&#97;&#115;&#99;&#64;&#108;&#117;&#109;&#99;&#46;&#110;&#108;</a>.
</p>

     



CHANGELOG.md
Changelog
==========

<!--

Newest changes should be on top.

This document is user facing. Please word the changes in such a way
that users understand how the changes affect the new version.
-->
version 1.3.0
---------------------
+ Added a workflow for somatic SV calling using GRIDSS, delly and manta.
+ Clever is now disabled by default. It can be enabled by setting `SVcalling.runClever` to `true`.
+ Fixed a bug which led to the removal of all BND, INS and INV events if DUPHOLD was enabled.
+ Updated SURVIVOR version to 1.0.7.
+ Added a workaround for a bug in survivor which caused the END of events detected by CLEVER to be calculated
  incorrectly resulting in incorrect merging of events. 
+ The gridss outputs are now included in survivor as well.
+ Smoove can now be disabled, by setting `SVcalling.runSmoove` to `false`.
+ Added Gridss as SV caller.
+ Fix order of input VCF files when removing missing and homozygous reference genotypes.
+ Each SV type has now seperate temporary directory during sorting.
+ Both union and intersection from each SV types are generated. 
+ The pipeline now first seperate SVs by types before merging.
+ Duphold can now be run optionally to annotate and remove FP deletions and duplications.
+ Replace excludeFpDupDel-option with runDuphold-option: DupHold annotation and FP filtering are now optional together.
+ Update CLEVER dockerimage.
+ DELLY: Update docker image.

version 1.2.0
---------------------
+ Structural-variantcalling pipeline: Remove GRIDSS from the pipeline.
+ Exclude GRIDSS from SURVIVOR merging: SVs were only defined as BNDs in GRIDSS.
+ Optional filtering of missing and hom-ref genotypes. 
+ Optional filtering of FP deletions and duplications.
+ Add DUPHOLD: annotate SVs with depth values.
+ Make bcftools indexing optional.
+ Structural-variantcalling pipeline: add sorting and change id.
+ Added GRIDSS sv caller

version 1.1.0
---------------------
+ Bcftools samtools and picard images are updated to newer versions.
+ Bcftools view now converts delly vcf to a gzipped version.
+ Tasks were updated to contain the `time_minutes` runtime attribute and
  associated `timeMinutes` input, describing the maximum time the task will
  take to run.

version 1.0.0
-----------------
+ Add WDL task for smoove SV-caller.
+ Created a structural variant-calling pipeline.


scripts/statistics.py
from cyvcf2 import VCF
import sys

def getBinSize(size):
    if size in range(50):
        return "BIN1: <50bp"
    elif size in range(50, 100):
        return "BIN2: 50-100bp"
    elif size in range(100, 1000):
        return "BIN3: 100-1000bp"
    elif size in range(1000, 10000):
        return "BIN4: 1000-10000bp"
    else:
        return "BIN5: >10000bp"


def getRecord(samples, sampleName, reader, vcf, output):
    with open(output, 'w') as FO:
        header = ",".join(["ID", "SAMPLE", "BINSIZE"] + list(samples))
        FO.write(header+"\n")
        for record in reader:
            id = record.CHROM+str(record.start)
            svtype = record.INFO["SVTYPE"]
            if svtype == "DEL":
                suppvec = record.INFO["SUPP_VEC"]
                size = record.INFO["AVGLEN"]
                binsize = getBinSize(abs(int(size)))
                data = ",".join([id, sampleName, binsize] + list(suppvec))
                FO.write(data+"\n")
                # print(data)


if __name__ == '__main__':
    vcf = sys.argv[1]
    sampleName = sys.argv[2]

    outputUpset = sys.argv[3]
    reader = VCF(vcf)
    samples = reader.samples
    getRecord(samples, sampleName, reader, vcf, outputUpset)


scripts/getUpSet.R
list.of.packages <- c("UpSetR")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages, repos = "https://cloud.r-project.org/")
library("UpSetR")

args <- commandArgs(trailingOnly = TRUE)
outFile <- args[3]
df <- read.csv(args[1], header=TRUE, sep=",")

bin1 <- df[df$BINSIZE %in% c("BIN1: <50bp"),]
bin2 <- df[df$BINSIZE %in% c("BIN2: 50-100bp"),]
bin3 <- df[df$BINSIZE %in% c("BIN3: 100-1000bp"),]
bin4 <- df[df$BINSIZE %in% c("BIN4: 1000-10000bp"),]
bin5 <- df[df$BINSIZE %in% c("BIN5: >10000bp"),]

pdf(outFile, paper = "a4")

upset(
    df, order.by = "freq",
    sets.bar.color="#56B4E9",
    empty.intersections = "on",
    text.scale = c(2, 1.3, 1, 1, 1.5, 2),
    point.size = 4,
    mainbar.y.label="All Size: Number of overlaps"
)

upset(
    bin1, order.by = "freq",
    sets.bar.color="#56B4E9",
    empty.intersections = "on",
    text.scale = c(2, 1.3, 1, 1, 1.5, 2),
    point.size = 4,
    mainbar.y.label="BIN1: <50bp: Number of overlaps"
)

upset(
    bin2, order.by = "freq",
    sets.bar.color="#56B4E9",
    empty.intersections = "on",
    text.scale = c(2, 1.3, 1, 1, 1.5, 2),
    point.size = 4,
    mainbar.y.label="BIN2: 50-100bp: Number of overlaps"
)

upset(
    bin3, order.by = "freq",
    sets.bar.color="#56B4E9",
    empty.intersections = "on",
    text.scale = c(2, 1.3, 1, 1, 1.5, 2),
    point.size = 4,
    mainbar.y.label="BIN3: 100-1000bp: Number of overlaps"
)

upset(
    bin4, order.by = "freq",
    sets.bar.color="#56B4E9",
    empty.intersections = "on",
    text.scale = c(2, 1.3, 1, 1, 1.5, 2),
    point.size = 4,
    mainbar.y.label="BIN4: 1000-10000bp: Number of overlaps"
)

upset(
    bin5, order.by = "freq",
    sets.bar.color="#56B4E9",
    empty.intersections = "on",
    text.scale = c(2, 1.3, 1, 1, 1.5, 2),
    point.size = 4,
    mainbar.y.label="BIN5: >10000bp: Number of overlaps"
)

#c(intersection size title, intersection size tick labels, set size title, set size tick labels, set names, numbers above bars)

dev.off()